src/remoted/secure.c

```diff
@@ -689,42 +711,123 @@ STATIC void HandleSecureMessage(const message_t *message, int *wdb_sock) {
+
+        minfo("Successfully reconnected to '%s'", DEFAULTQUEUE);
+
         if (SendMSG(logr.m_queue, tmp_msg, srcmsg, SECURE_MQ) < 0) {
+            // Something went wrong sending a message after an immediate reconnection...
             merror(QUEUE_ERROR, DEFAULTQUEUE, strerror(errno));
+        } else {
+            rem_inc_recv_evt(agentid_str);
+        }
+    } else {
+        rem_inc_recv_evt(agentid_str);
+    }
 
-            // Try to reconnect infinitely
-            logr.m_queue = StartMQ(DEFAULTQUEUE, WRITE, INFINITE_OPENQ_ATTEMPTS);
+    router_message_forward(tmp_msg, keys.keyentries[agentid]->id);
 
-            minfo("Successfully reconnected to '%s'", DEFAULTQUEUE);
+    os_free(agentid_str);
+}
 
-            if (SendMSG(logr.m_queue, tmp_msg, srcmsg, SECURE_MQ) < 0) {
-                // Something went wrong sending a message after an immediate reconnection...
-                merror(QUEUE_ERROR, DEFAULTQUEUE, strerror(errno));
-            } else {
-                rem_inc_recv_evt(agentid_str);
+void router_message_forward(char* msg, const char* agent_id) {
+    // Both syscollector delta and sync messages are sent to the router
+    ROUTER_PROVIDER_HANDLE router_handle = NULL;
+    int message_header_size = 0;
+    flatcc_json_parser_table_f *parser = NULL;
+
+    if(strncmp(msg, SYSCOLLECTOR_HEADER, SYSCOLLECTOR_HEADER_SIZE) == 0) {
+        w_mutex_lock(&router_syscollector_mutex);
+        if (!router_syscollector_handle) {
+            if (router_syscollector_handle = router_provider_create("syscollector"), !router_syscollector_handle) {
+                mdebug2("Failed to create router handle for 'syscollector'.");
+                w_mutex_unlock(&router_syscollector_mutex);
+                return;
+            }
+        }
+        router_handle = router_syscollector_handle;
+        message_header_size = SYSCOLLECTOR_HEADER_SIZE;
+        parser = Syscollector_Delta_parse_json_table;
+        w_mutex_unlock(&router_syscollector_mutex);
+    } else if(strncmp(msg, DBSYNC_SYSCOLLECTOR_HEADER, DBSYNC_SYSCOLLECTOR_HEADER_SIZE) == 0) {
+        w_mutex_lock(&router_rsync_mutex);
+        if (!router_rsync_handle) {
+            if (router_rsync_handle = router_provider_create("rsync"), !router_rsync_handle) {
+                mdebug2("Failed to create router handle for 'rsync'.");
+                w_mutex_unlock(&router_rsync_mutex);
+                return;
             }
+        }
+        router_handle = router_rsync_handle;
+        message_header_size = DBSYNC_SYSCOLLECTOR_HEADER_SIZE;
+        parser = Syscollector_SyncMsg_parse_json_table;
+        w_mutex_unlock(&router_rsync_mutex);
+    }
+
+    if (!router_handle) {
+        return;
+    }
+
+    cJSON* j_msg_to_send = NULL;
+    cJSON* j_agent_info = NULL;
+    cJSON* j_msg = NULL;
+    char* msg_to_send = NULL;
+
+    char* msg_start = msg + message_header_size;
+    size_t msg_size = strnlen(msg_start, OS_MAXSTR);
```

Comment:
Size limit for strnlen should be "OS_MAXSTR - message_header_size"

Reply:
You are right, thank you!